homeassistant/helpers/config_entry_oauth2_flow.py

```diff
@@ -233,6 +242,80 @@ async def _token_request(self, data: dict) -> dict:
         return cast(dict, await resp.json())
 
 
+class LocalOAuth2ImplementationWithPkce(LocalOAuth2Implementation):
+    """Local OAuth2 implementation with PKCE."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        domain: str,
+        client_id: str,
+        authorize_url: str,
+        token_url: str,
+        client_secret: str = "",
+        code_verifier_length: int = 100,
+    ) -> None:
+        """Initialize local auth implementation."""
+        super().__init__(
+            hass,
+            domain,
+            client_id,
+            client_secret,
+            authorize_url,
+            token_url,
+        )
+
+        # Generate code verifier
+        self.code_verifier = LocalOAuth2ImplementationWithPkce.generate_code_verifier(
+            code_verifier_length
+        )
+
+    @property
+    def code_challenge_data(self) -> dict:
+        """Return the generated code challenge and method."""
+        return {
+            "code_challenge": LocalOAuth2ImplementationWithPkce.compute_code_challenge(
+                self.code_verifier
+            ),
+            "code_challenge_method": "S256",
+        }
+
+    @property
+    def extra_authorize_data(self) -> dict:
+        """Extra data that needs to be appended to the authorize url."""
+        return self.code_challenge_data
+
+    @property
+    def code_verifier_data(self) -> dict:
+        """Return the code verifier."""
+        return {"code_verifier": self.code_verifier}
+
+    @property
+    def extra_token_redeem_data(self) -> dict:
+        """Extra data that needs to be included in the token redeem request."""
+        return self.code_verifier_data
+
+    @staticmethod
+    def generate_code_verifier(length: int = 100) -> str:
+        """Generate a code verifier."""
+        code_verifier = base64.urlsafe_b64encode(os.urandom(length + 20)).decode(
```

Comment:
Probably use the `secrets` module to generate tokens that should be hard to guess:
https://docs.python.org/3/library/secrets.html#secrets.token_urlsafe

https://www.rfc-editor.org/rfc/rfc7636#section-4.1